add plugin nearest exon junction boundary

ima23 opened the pull request:
Plugin to report nearest exon junction boundary for variants in the protein coding region.
Reports:
- exon identifier
- bp distance
- start /end if it's nearest to the start/end of the exon
- exon bp lenght

aparton commented:
Hiya,

Comments below. Some are more valid than others, so feel free to reject:

- Do we want it to return the nearest boundary for a variant even if it's intronic? Admittedly David is only looking at coding variants, but I think it could be a useful for others
- $loc_string isn't used, it can be removed
- I can't make it display multiple boundaries if they're equidistant like it says in the description - we could easily just prioritise end ones
- With the offline flag on, I throw an error at Line 107 before the warning on line 109
- Adding the plugin name to the error message on 109 allows would allow users to identify which plugin isn't working if they're running lots at teh same time
- Might it be worth adding 'BP' to the length result to make it easier to read?
- Might it be worth adding +/- to the distance to indicate which direction the boundary is in?
- Can we add the offline warning to the text description?
- Typo in the description: "the distance of the to the exon boundary"

ima23 commented:
Thanks for the comments @aparton!
> Comments below. Some are more valid than others, so feel free to reject:
> 
> * Do we want it to return the nearest boundary for a variant even if it's intronic? Admittedly David is only looking at coding variants, but I think it could be a useful for others

I think this is debatable. Can make another PR with the NearestExon.pm plugin which covers this case as it reports the nearest exon overall and not in respect to a particular transcript as this plugin.

> * $loc_string isn't used, it can be removed

Well spotted. Thanks.

> * I can't make it display multiple boundaries if they're equidistant like it says in the description - we could easily just prioritise end ones

Can you give me please the example that did not work? Thanks.

> * With the offline flag on, I throw an error at Line 107 before the warning on line 109

Yep, updated that.

> * Adding the plugin name to the error message on 109 allows would allow users to identify which plugin isn't working if they're running lots at the same time

Can you give me an example please? In the case of L84 and and L101, the errors automatically get the plugin name in the message.

> * Might it be worth adding 'BP' to the length result to make it easier to read?

It might be easier human readable, however I expect the users will parse the output files more often the open them manually, in which case using the length value would require an extra split/match.

> * Might it be worth adding +/- to the distance to indicate which direction the boundary is in?

Can you please give me an example of what you mean? Thanks.

> * Can we add the offline warning to the text description?

Yep, added.

> * Typo in the description: "the distance of the to the exon boundary"

Well spotted. Thanks.

aparton commented:
Hiya,

Thanks for these. It's pretty much there, I only have one issue: 

- Regarding equidistant variants, I've got it up and running now, it wasn't handling insertions/deletions the way I expected. For example, if I have a SNP:

perl vep --id 'X 123480502 rsTest A T' --database --force_overwrite --db_version 95 --port 3337 --Plugin NearestExonJB 

Then I get the output: ENSE00001454450|127|start_end|255

Or if I use the deletion: 

perl vep --id 'X 123480501 rsTest TAA T' --database --force_overwrite --db_version 95 --port 3337 --Plugin NearestExonJB 

Then I get the same output: ENSE00001454450|127|start_end|255 - should the second one not be 126 away?




Follow-ups to your previous comments:

- Regarding handing intronic variants, I think it would be useful but I take your point that it doesn't need to be part of this plugin and we could create another one/add it later.
- Offline flag is now reaching the error message, and the plugin name is in there
- Fair point about not adding BP. You're right, they're more likely to parse the output. Regarding adding +/- to the distance, I meant something like ENSE00001946082|**-**63|start|169 or ENSE00001946082|**+**63|end|169 to indicate whether you move 5' or 3' to the nearest junction boundary, but the same argument about just adding something else to parse applies here as well.

willmclaren commented:
Sorry to be a PITA from beyond the Ensembl grave, but you might want to make use of the methods already in place on the TranscriptVariation object:

https://github.com/Ensembl/ensembl-variation/blob/release/95/modules/Bio/EnsEMBL/Variation/BaseTranscriptVariation.pm#L572-L661

These are already super-optimised for finding the exon(s) overlapping the variant. Their return values are cached and will already have been pre-computed by the time the plugin runs. This avoids fetching adaptors and all that nonsense too.

Do you also need to account for the case where a variant overlaps more than one exon (for example a long-ish deletion)?

ima23 commented:
Thanks for the example @aparton,  yes, the result is the same as the normalised position of both inputs is 123480502, leading to the same distance to start/end of exon.

> perl vep --id 'X 123480502 rsTest A T' --database --force_overwrite --db_version 95 --port 3337 --Plugin NearestExonJB

Thanks for pointing out the code @willmclaren. I agree, the more optimised code is used the better, will look into the methods. The current use case is aiming more to cover smaller variants than larger structural variants, and within one transcript there will be most of the time only one nearest exon junction boundary. However, if you have a certain code section in mind/ use example, I'd be interested to know.

I will review the code in reflection of the cached methods/data.

willmclaren commented on NearestExonJB.pm:
```diff
@@ -105,20 +103,18 @@ sub run {
 
-    my @exons = @{$trv->transcript->get_all_Exons};
+  if(!exists($self->{_cache}) || !exists($self->{_cache}->{$loc_string})) {
+    my @exons = @{$trv->_exons};
```
You don't need to access exon number to get the overlapped exon.

I'd suggest instead:

```
my @exons = $trv->_overlapped_exons;

if(scalar @exons == 1) {
  $exon = $exons[0];
  ## rest of code as you have it
}
else {
  ## either no exon overlap (@exons == 0) or overlaps more than 1 (@exons > 1)
  ## bail out and return {} ???
}
```
